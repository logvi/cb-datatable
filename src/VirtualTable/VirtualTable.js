// @flow
import React, {Component} from 'react';
import FixedHeaderTable from '../FixedHeaderTable/FixedHeaderTable';
import type VirtualTableProps from './VirtualTable.type';

class VirtualTable extends Component {
  props: VirtualTableProps;
  tableComponent: FixedHeaderTable;
  transformY: number;
  tableLayout: Function;
  scrollArea: HTMLElement;
  fullArea: HTMLElement;
  tableBody: HTMLElement;
  tableBodyScrollHandler: MouseEventListener;

  constructor(props: VirtualTableProps) {
    super(props);
    this.tableLayout = this.tableLayout.bind(this);

    this.transformY = 0;
  }

  componentDidMount() {
    this.scrollArea.addEventListener('scroll', this.tableBodyScrollHandler);
    // set full area height by data length
    this.setFullHeight(this.props.data.length * this.avgRowHeight);
    this.initRows();
  }

  componentDidUpdate() {
    this.transformY = this.props.startIndex * this.avgRowHeight;
    // if height of loaded table + offset top more than full container height, update full container height
    const innerTableHeight = this.transformY + this.tableComponent.tableBodyContainer.offsetHeight;
    if (
      innerTableHeight > this.fullArea.offsetHeight ||
      (this.props.data.length === this.props.endIndex + 1 &&
      this.fullArea.offsetHeight > innerTableHeight)
    )
    {
      this.setFullHeight(innerTableHeight);
    }

    // apply transform for table
    this.tableComponent.tableBodyContainer.style.transform = `translate3d(0, ${this.transformY}px, 0)`;
  }

  get avgRowHeight(): number {
    const tableBodyHeight = this.tableComponent.tableBody.offsetHeight;
    return tableBodyHeight / this.props.displayData.length;
  }

  initRows() {
    if (this.scrollArea.offsetHeight > this.tableComponent.tableBodyContainer.offsetHeight) {
      const displayedEndIndex = Math.floor(this.scrollArea.offsetHeight / this.avgRowHeight);
      this.props.setIndex(0, displayedEndIndex + 2);
    }
  }

  setFullHeight(value: number) {
    this.fullArea.style.height = `${value}px`;
    this.fullArea.style.maxHeight = `${value}px`;
  }

  tableBodyScrollHandler = (e: EventHandler & {target: HTMLElement}) => {
    // scroll horizontal
    const {target} = e;
    const {scrollLeft} = target;
    this.tableComponent.tableHeaderContainer.style.transform = 'translate3d(' + (-1 * scrollLeft) + 'px, 0, 0)';

    // scroll vertical
    if (target.scrollTop + this.scrollArea.offsetHeight > this.tableComponent.tableBodyContainer.offsetHeight + this.transformY) {
      // load down
      const displayedStartIndex = Math.floor(target.scrollTop / this.avgRowHeight);
      const displayedEndIndex = displayedStartIndex + Math.floor(this.scrollArea.offsetHeight / this.avgRowHeight);
      this.props.setIndex(displayedStartIndex - 2, displayedEndIndex + 2);
    } else if (target.scrollTop < this.transformY) {
      // load up
      const displayedStartIndex = Math.floor(target.scrollTop / this.avgRowHeight);
      const displayedEndIndex = displayedStartIndex + Math.floor(this.scrollArea.offsetHeight / this.avgRowHeight);
      this.props.setIndex(displayedStartIndex - 2, displayedEndIndex + 2);
    }
  };

  tableLayout({Body, Header, Loader}: {Body: React$Element<*>, Header: React$Element<*>, Loader: React$Element<*>}) {
    return (
      <div className="fixed-header-table virtual-data-table">
        {Header}
        {Loader}
        <div className="virtual-data-table__scroll-area" ref={elem => this.scrollArea = elem}>
          <div className="virtual-data-table__full-area" ref={elem => this.fullArea = elem}>
            {Body}
          </div>
        </div>
      </div>
    );
  }

  render() {
    return (
      <FixedHeaderTable
        {...this.props}
        renderTableLayout={this.tableLayout}
        ref={el => this.tableComponent = el}
      />
    );
  }
}

export default VirtualTable;
