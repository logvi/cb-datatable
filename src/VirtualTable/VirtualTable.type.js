import type {TableProps} from '../types';

type VirtualTableProps = TableProps & {
  startIndex: number,
  endIndex: number,
  setIndex: Function
};

export default VirtualTableProps;
