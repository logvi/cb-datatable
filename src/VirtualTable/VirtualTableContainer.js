import DataTableContainer from '../DataTableContainer';
import VirtualTable from './VirtualTable';

class VirtualTableContainer extends DataTableContainer {
  static defaultProps = {
    ...DataTableContainer.defaultProps,
    TableComponent: VirtualTable,
    PagerComponent: null
  };
}

export default VirtualTableContainer;
