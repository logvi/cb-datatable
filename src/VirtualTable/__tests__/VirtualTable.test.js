import VirtualTable from '../';
import GeneralTableTests from '../../__tests__/GeneralTableTests';

describe('VirtualTable', GeneralTableTests(VirtualTable));
