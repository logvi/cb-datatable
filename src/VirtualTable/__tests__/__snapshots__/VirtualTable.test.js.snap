// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VirtualTable Columns visibility. Column 1 should be hidden 1`] = `
<div
  className="cb-datatable"
  onClick={undefined}
>
  <div
    className="fixed-header-table virtual-data-table"
  >
    <div
      className="fixed-header-table__header"
    >
      <table>
        <thead>
          <tr>
            <th
              className="sortable"
              onClick={[Function]}
            >
              <div
                className="th-content"
              >
                column0
              </div>
            </th>
            <th
              className="sortable"
              onClick={[Function]}
            >
              <div
                className="th-content"
              >
                column2
              </div>
            </th>
            <th
              className="sortable"
              onClick={[Function]}
            >
              <div
                className="th-content"
              >
                column3
              </div>
            </th>
            <th
              className="sortable"
              onClick={[Function]}
            >
              <div
                className="th-content"
              >
                column4
              </div>
            </th>
          </tr>
        </thead>
      </table>
    </div>
    <div
      className="virtual-data-table__scroll-area"
    >
      <div
        className="virtual-data-table__full-area"
      >
        <div
          className="fixed-header-table__body"
        >
          <table>
            <tbody>
              <tr
                onClick={null}
              >
                <td
                  className={null}
                >
                  <div
                    className="table-cell-content"
                  >
                    column0 0
                  </div>
                </td>
                <td
                  className={null}
                >
                  <div
                    className="table-cell-content"
                  >
                    column2 0
                  </div>
                </td>
                <td
                  className={null}
                >
                  <div
                    className="table-cell-content"
                  >
                    column3 0
                  </div>
                </td>
                <td
                  className={null}
                >
                  <div
                    className="table-cell-content"
                  >
                    column4 0
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`VirtualTable orderBy asc Snapshot: should apply sort class and icon 1`] = `
<div
  className="cb-datatable"
  onClick={undefined}
>
  <div
    className="fixed-header-table virtual-data-table"
  >
    <div
      className="fixed-header-table__header"
    >
      <table>
        <thead>
          <tr>
            <th
              className="sortable sort-asc"
              onClick={[Function]}
            >
              <div
                className="th-content"
              >
                <div
                  className="th-content__label"
                >
                  column 0
                </div>
                <span
                  className="icon-sort icon-sort-asc"
                >
                   ▲
                </span>
              </div>
            </th>
          </tr>
        </thead>
      </table>
    </div>
    <div
      className="virtual-data-table__scroll-area"
    >
      <div
        className="virtual-data-table__full-area"
      >
        <div
          className="fixed-header-table__body"
        >
          <table>
            <tbody>
              <tr
                onClick={null}
              >
                <td
                  className={null}
                >
                  <div
                    className="table-cell-content"
                  >
                    column0 0
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`VirtualTable orderBy desc Snapshot: should apply sort class and icon 1`] = `
<div
  className="cb-datatable"
  onClick={undefined}
>
  <div
    className="fixed-header-table virtual-data-table"
  >
    <div
      className="fixed-header-table__header"
    >
      <table>
        <thead>
          <tr>
            <th
              className="sortable sort-desc"
              onClick={[Function]}
            >
              <div
                className="th-content"
              >
                <div
                  className="th-content__label"
                >
                  column 0
                </div>
                <span
                  className="icon-sort icon-sort-desc"
                >
                   ▼
                </span>
              </div>
            </th>
          </tr>
        </thead>
      </table>
    </div>
    <div
      className="virtual-data-table__scroll-area"
    >
      <div
        className="virtual-data-table__full-area"
      >
        <div
          className="fixed-header-table__body"
        >
          <table>
            <tbody>
              <tr
                onClick={null}
              >
                <td
                  className={null}
                >
                  <div
                    className="table-cell-content"
                  >
                    column0 2
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`VirtualTable should apply renderLayout prop, it displays only table 1`] = `
<div
  className="fixed-header-table virtual-data-table"
>
  <div
    className="fixed-header-table__header"
  >
    <table>
      <thead>
        <tr>
          <th
            className="sortable"
            onClick={[Function]}
          >
            <div
              className="th-content"
            >
              column 0
            </div>
          </th>
        </tr>
      </thead>
    </table>
  </div>
  <div
    className="virtual-data-table__scroll-area"
  >
    <div
      className="virtual-data-table__full-area"
    >
      <div
        className="fixed-header-table__body"
      >
        <table>
          <tbody>
            <tr
              onClick={null}
            >
              <td
                className={null}
              >
                <div
                  className="table-cell-content"
                >
                  column0 0
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
